### PR TITLE
Add some basic guidelines around using C++ standard headers now that our build allows us to

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Just putting some basic guidelines in our docs as a starting point.

### CODE REVIEW

File: docs/coding-guidelines/clr-code-guide.md
Comment (AaronRobinsonMSFT) [inline docs/coding-guidelines/clr-code-guide.md:1261]:
```suggestion
The C and C++ standard headers are available for usage in the CoreCLR code-base. However, there are a restrictions on using the standard-provided APIs for code that will run as part of CoreCLR.
```

File: docs/coding-guidelines/clr-code-guide.md
Comment (AaronRobinsonMSFT) [inline docs/coding-guidelines/clr-code-guide.md:1263]:
```suggestion
Code that will only run in other processes, such as `createdump` or other extraneous tools, do not have the same set of restrictions.
```

File: docs/coding-guidelines/clr-code-guide.md
Comment (AaronRobinsonMSFT) [inline docs/coding-guidelines/clr-code-guide.md:1267]: Perhaps add a link to the top level directory for the PAL and minipal?

File: docs/coding-guidelines/clr-code-guide.md
Comment (AaronRobinsonMSFT) [inline docs/coding-guidelines/clr-code-guide.md:1285]: Reference or link to section `2.11` for more details?

File: docs/coding-guidelines/clr-code-guide.md
Comment (jkotas) [inline docs/coding-guidelines/clr-code-guide.md:1271]: On Linux, we build against ROOTFS generated from Linux versions that is out of support. We have convinced ourselves that it is ok given that we are using just prototypes and no actual code from ROOTFS.

Do C++ standard headers come from the ROOTFS? C++ headers do contain actual code (e.g. collection types). If we start including C++ headers with actual code from ROOTFS, our reasoning for why the ROOTFS is ok falls apart.

Comment (jkoritzinsky) [reply docs/coding-guidelines/clr-code-guide.md:1271]: We do use some code from the ROOTFS today. We get some object files linked in from the relevant libc in the ROOTFS. This does give us more exposure to the C++ standard library version from the ROOTFS though, which is worth considering with respect to security issues.

File: docs/coding-guidelines/clr-code-guide.md
Comment (danmoseley) [reply docs/coding-guidelines/clr-code-guide.md:1261]: "a restrictions" .. still a typo

File: docs/coding-guidelines/clr-code-guide.md
Comment (am11) [reply docs/coding-guidelines/clr-code-guide.md:1267]: Should we mention `unsigned char` used by non-win32 in minipal/utf8 internals and usages: https://github.com/dotnet/runtime/blob/6fc9636509d014af530add71a6186218be90df55/src/native/minipal/utf8.h#L21-L25

Comment (am11) [reply docs/coding-guidelines/clr-code-guide.md:1267]: Should we mention `unsigned short` used by non-win32 in minipal/utf8 internals and usages: https://github.com/dotnet/runtime/blob/6fc9636509d014af530add71a6186218be90df55/src/native/minipal/utf8.h#L21-L25

File: docs/coding-guidelines/clr-code-guide.md
Comment (jkoritzinsky) [reply docs/coding-guidelines/clr-code-guide.md:1271]: I've been investigating this further based on our offline conversations, and I've noticed that tools like createdump are already using C++ headers with the template types. I'd assume that we'd have the same concern there as well as in the hosting layer (which uses C++ standard collections and other template types as well).

Comment (jkoritzinsky) [reply docs/coding-guidelines/clr-code-guide.md:1271]: https://github.com/dotnet/arcade/pull/14728 is the first step towards ensuring we're using maintained/supported C++ standard headers.

**Review by AaronRobinsonMSFT — APPROVED** [review]
LGTM as long as @jkotas's concerns about building from a ROOTFS are addressed.

File: docs/coding-guidelines/clr-code-guide.md
Comment (AaronRobinsonMSFT) [inline docs/coding-guidelines/clr-code-guide.md:1265]:
```suggestion
To ensure we're using a supported and easily updatable standard library implementation, we use a specially built libc++ implementation, on non-Windows, in our shipping products. This enables us to use a modern C++ standard library implementation while still targeting older rootfs targets with outdated C++ library implementations.
```

Where are we getting this special build?

Comment (am11) [reply docs/coding-guidelines/clr-code-guide.md:1265]: In rootfs, e.g. https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/0b9f68c13626e2e15d916d1565191501b2e0efc5/src/azurelinux/3.0/net9.0/cross/amd64/Dockerfile#L4, xenial (Ubuntu 16.04) has libc 2.23 (current version is 2.39). So we are essentially exercising the backward-compatibility (unwritten) rule of glibc; link with older libc to maximize the compat. This does not stop us from using the latest _toolchain_ however; case in point, we are using the latest-ish clang in our builds with `-sysroot $ROOTFS_DIR`. 🚀

Comment (jkoritzinsky) [reply docs/coding-guidelines/clr-code-guide.md:1265]:
```suggestion
To ensure we're using a supported and easily updatable standard library implementation, we use a specially built libc++ implementation, in our Linux Docker containers, in our shipping products. This enables us to use a modern C++ standard library implementation while still targeting older rootfs targets with outdated C++ library implementations.
```

File: docs/coding-guidelines/clr-code-guide.md
Comment (jkotas) [inline docs/coding-guidelines/clr-code-guide.md:1273]: > as long as only the non-throwing members are used

What is an example of throwing member that cannot be used?

Comment (huoyaoyuan) [reply docs/coding-guidelines/clr-code-guide.md:1273]: Is `bad_alloc` something new or existed before? Can STL containers be used more to replace custom containers?

Comment (AaronRobinsonMSFT) [reply docs/coding-guidelines/clr-code-guide.md:1273]: > I'll add it as an example (and mention to look at cppreference.com or the C++ standard to check the contracts for members).

This is not going to be sufficient. There are many more concerns with these collections and their use in the coreclr codebase. I appreciate we are starting to document this, but the `bad_alloc` case is the tip of the iceberg here. We have no support in our system's contracts, these have caught innumerable bugs at dev time, so we need a conversation on that. Functions that are marked `noexcept` and encounter an exception will now result in a C++ run-time's "fail fast" occurring - this would be very new behavior that must be carefully considered across all standard implementations since currently our contracts would find it or it could be caught by some `EX_TRY`.

My point here is this guidance will need a lot more consideration until we use any C++ collection. Introduction of this kind of use in the .NET 9 release should be avoided or only in the most trivial and appropriate cases. We've a lot of learn in this area and accomodate in the current code base. My recommendation is we approach this cautiously.

> Can STL containers be used more to replace custom containers?

This is going to be a long term goal and not something we are likely to accept from community members until the core team develops more guidelines on use. Really appreciate the enthusiasm and look forward to help in this area, but for .NET 9 and likely .NET 10 our use of these collections is going to be narrow and very limited.

**Comment by jkoritzinsky** [comment]
Circling back to this: Trying to use libc++ with libstdc++ as the ABI doesn't work for Ubuntu 16.04. As a result, I've updated the guidance to recommend against template types for now. We may revisit these rules in .NET 10, but these rules should be a good starting point.

**Review by jkotas — APPROVED** [review]

**Pull request merged**
